lib/util/cmcd_manager.js

```diff
@@ -0,0 +1,761 @@
+    data.ot = this.getObjectTypeFromMimeType_(mimeType);
+    data.su = true;
+
+    const query = shaka.util.CmcdManager.toQuery(data);
+
+    return shaka.util.CmcdManager.appendQueryToUri(uri, query);
+  }
+
+  /**
+   * Apply CMCD data to side car text track uri.
+   *
+   * @param {string} uri
+   */
+  appendTextTrackData(uri) {
+    if (!this.config_.enabled) {
+      return uri;
+    }
+
+    const data = this.createData_();
+    data.ot = shaka.util.CmcdManager.ObjectType.CAPTION;
+    data.su = true;
+
+    const query = shaka.util.CmcdManager.toQuery(data);
+
+    return shaka.util.CmcdManager.appendQueryToUri(uri, query);
+  }
+
+  /**
+   * The CMCD object type.
+   *
+   * @param {!shaka.util.CmcdManager.SegmentInfo} segment
+   * @return {(shaka.util.CmcdManager.ObjectType|undefined)}
+   * @private
+   */
+  getObjectType_(segment) {
+    const type = segment.stream.type;
+
+    if (type == 'text' && segment.stream.mimeType === 'text/vtt') {
+      return shaka.util.CmcdManager.ObjectType.CAPTION;
+    }
+
+    if (segment.init) {
+      return shaka.util.CmcdManager.ObjectType.INIT;
+    }
+
+    if (type == 'video') {
+      return shaka.util.CmcdManager.ObjectType.VIDEO;
+    }
+
+    if (type == 'audio') {
+      return shaka.util.CmcdManager.ObjectType.AUDIO;
+    }
+
+    if (type == 'text') {
+      if (segment.stream.mimeType === 'application/mp4') {
+        return shaka.util.CmcdManager.ObjectType.TIMED_TEXT;
+      }
+      return shaka.util.CmcdManager.ObjectType.CAPTION;
+    }
+
+    return undefined;
+  }
+
+  /**
+   * The CMCD object type from mimeType.
+   *
+   * @param {!string} mimeType
+   * @return {(shaka.util.CmcdManager.ObjectType|undefined)}
+   * @private
+   */
+  getObjectTypeFromMimeType_(mimeType) {
+    switch (mimeType) {
+      case 'video/webm':
+      case 'video/mp4':
+        return shaka.util.CmcdManager.ObjectType.MUXED;
+
+      case 'application/x-mpegurl':
+        return shaka.util.CmcdManager.ObjectType.MANIFEST;
+
+      default:
+        return undefined;
+    }
+  }
+
+  /**
+   * Get the buffer length for a media type in milliseconds
+   *
+   * @param {string} type
+   * @return {number}
+   * @private
+   */
+  getBufferLength_(type) {
+    const ranges = this.playerInterface_.getBufferedInfo()[type];
+    let start = 0;
+    let end = 0;
+
+    for (const range of ranges) {
+      if (range.start < start) {
+        start = range.start;
+      }
+      if (range.end > end) {
+        end = range.end;
+      }
+    }
+
+    return (end - start) * 1000;
+  }
+
+  /**
+   * Get the stream type
+   *
+   * @return {shaka.util.CmcdManager.StreamType}
+   */
+  getStreamType() {
+    const isLive = this.playerInterface_.isLive();
+    if (isLive) {
+      return shaka.util.CmcdManager.StreamType.LIVE;
+    } else {
+      return shaka.util.CmcdManager.StreamType.VOD;
+    }
+  }
+
+  /**
+   * Generate a random v4 UUID
+   *
+   * @return {string}
+   */
+  static uuid() {
```

Comment:
One idea is to use the native function of modern browsers, I suggest using:
```
if (window.crypto && 'randomUUID' in window.crypto) {
  return window.crypto.randomUUID();
}
```

Reply:
That function is only available in recent browser versions. How would we support platforms like Smart TVs that are locked to older versions?

Reply:
Something like this would be enough:
```
static uuid() {
  if (window.crypto && 'randomUUID' in window.crypto) {
    return window.crypto.randomUUID();
  }
  const url = URL.createObjectURL(new Blob());
  const uuid = url.toString();
  URL.revokeObjectURL(url);
  return uuid.substr(uuid.lastIndexOf('/') + 1);
}
```

Reply:
More info: https://github.com/WICG/uuid

Reply:
If the intent is to alway use standard W3C methods, wouldn't it be better to polyfill the function instead adding a `null` checked utility function?

Reply:
@joeyparrish what do you think?

Reply:
Clearly, Shaka Player [loves polyfills](https://github.com/google/shaka-player/tree/master/lib/polyfill), though [not everyone agrees with this](https://github.com/google/shaka-player/issues/2625).

If you want to polyfill `crypto.randomUUID`, I think I would be very much okay with that.  However, if the only use of it is in a single static utility function, then I think it's just as good to handle any necessary fallback in the function itself and skip polyfills.

Reply:
I double-checked the spec for createObjectURL, and it is specified that the URL must use a random UUID.  So this should be safe everywhere.

Reply:
I created a polyfill for it, see: https://github.com/google/shaka-player/pull/3669

Reply:
Once this is merged I will update the code to use `crypto.randomUUID` directly.